### PR TITLE
Added possibility to deploy a HAProxy as an enhanced MQTT load balancer

### PR DESCRIPTION
**Pull request by mattkaem**
Added possibility to deploy a HAProxy as an enhanced MQTT load balancer and added a section on how to customize the deployment in the README

### CODE REVIEW

File: terraform/modules/hono/variables.tf [hunk 1]
```diff
@@ -207,3 +207,15 @@ variable "grafana_dns_name" {
   type        = string
   description = "Name of the DNS host for Grafana"
 }
+
+variable "enhanced_mqtt_load_balancer" {
```
Comment (julian-sotec) [inline terraform/modules/hono/variables.tf:211]: Rename to advanced_load_ablancer - when moved under mqtt_adapter object

File: terraform/modules/hono/README.md [hunk 2]
```diff
@@ -1,3 +1,4 @@
+<!-- BEGIN_TF_DOCS -->
```
Comment (julian-sotec) [inline terraform/modules/hono/README.md:1]: Please remove if possible

File: terraform/modules/hono/README.md [hunk 3]
```diff
@@ -75,3 +77,4 @@ No modules.
 | Name | Description |
 |------|-------------|
 | <a name="output_values"></a> [values](#output\_values) | n/a |
+<!-- END_TF_DOCS -->
```
Comment (julian-sotec) [inline terraform/modules/hono/README.md:80]: Please remove if possible

File: terraform/modules/load_balancer/main.tf [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        replicaCount = var.enhanced_mqtt_load_balancer.replicaCount
+        service = {
+          tcpPorts = [
+            for port_config in var.enhanced_mqtt_load_balancer.port_configs :
+            port_config
+          ]
+          type: "LoadBalancer"
+          enablePorts = {
+            http: false
+            https: false
```
Comment (julian-sotec) [inline terraform/modules/load_balancer/main.tf:14]: When two mqtt ports are specified e.g: 8883 and 443 - will https = false disable mqtt traffic on port 443?

File: terraform/modules/load_balancer/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+| [kubernetes_config_map.tcp](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/config_map) | resource |
+
+## Inputs
+
+| Name | Description | Type | Default | Required |
+|------|-------------|------|---------|:--------:|
+| <a name="input_enable_mqtt_adapter"></a> [enable\_mqtt\_adapter](#input\_enable\_mqtt\_adapter) | Used to enable the MQTT adapter. | `bool` | n/a | yes |
+| <a name="input_enhanced_mqtt_load_balancer"></a> [enhanced\_mqtt\_load\_balancer](#input\_enhanced\_mqtt\_load\_balancer) | Configuration options for the enhanced MQTT load balancer.<br>  chart\_version: Version of the chart to deploy.<br>  replicaCount: Number of replicas to deploy.<br>  port\_configs: List of MQTT port config objects for the enhanced MQTT load balancer service.<br>  tcp\_configmap\_data: Data of the TCP configMap for the enhanced MQTT load balancer. | <pre>object({<br>    chart_version = string<br>    replicaCount = number<br>    port_configs = list(object({<br>      name       = string,<br>      port       = number,<br>      targetPort = number<br>    }))<br>    tcp_configmap_data = map(string)<br>  })</pre> | n/a | yes |
```
Comment (julian-sotec) [inline terraform/modules/load_balancer/README.md:28]: can we document the specific settings of the load balancer in an additional table and reference the lb object here with a minimum description?

File: terraform/software/README.md [hunk 6]
```diff
@@ -1,3 +1,4 @@
+<!-- BEGIN_TF_DOCS -->
```
Comment (julian-sotec) [inline terraform/software/README.md:1]: Please remove

File: terraform/software/README.md [hunk 7]
```diff
@@ -86,3 +89,4 @@ No resources.
 | <a name="output_hono_tls_crt_in_storage"></a> [hono\_tls\_crt\_in\_storage](#output\_hono\_tls\_crt\_in\_storage) | n/a |
 | <a name="output_hono_tls_key_in_storage"></a> [hono\_tls\_key\_in\_storage](#output\_hono\_tls\_key\_in\_storage) | n/a |
 | <a name="output_values"></a> [values](#output\_values) | n/a |
+<!-- END_TF_DOCS -->
```
Comment (julian-sotec) [inline terraform/software/README.md:92]: Please remove

File: terraform/software/README.md [hunk 8]
```diff
@@ -40,6 +42,7 @@ No resources.
...
...T load balancer.<br>  chart\_version: Version of the chart to deploy.<br>  algorithm: Load balancing algorithm used by the enhanced MQTT load balancer. For a list of possible options see https://www.haproxy.com/documentation/kubernetes-ingress/community/configuration-reference/ingress/#load-balance .<br>  replicaCount: Number of replicas to deploy.<br>  port\_configs: List of MQTT port config objects for the enhanced MQTT load balancer service.<br>  tcp\_configmap\_data: Data of the TCP configMap for the enhanced MQTT load balancer. | <pre>object({<br>    enabled = optional(bool, true),<br>    chart_version = optional(string, "1.34.1"),<br>    algorithm = optional(string, "leastconn"),<br>    replicaCount = optional(number, 1),<br>    port_configs = optional(list(object({<br>      name       = string,<br>      port       = number,<br>      targetPort = optional(number, 8883)<br>    })), [<br>            {<br>              name: "mqtt"<br>              port: 8883<br>              targetPort: 8883<br>            }<br>          ]),<br>    tcp_configmap_data = optional(map(string), {<br>            8883 = "hono/eclipse-hono-adapter-mqtt:8883"<br>          })<br>  })</pre> | `{}` | no |
```
Comment (julian-sotec) [inline terraform/software/README.md:45]: The description including the link (explaination) should be moved out of the table if possible. Can we maybe create a new section with a table where we describe it? link should be in a description outside the table if possible

File: terraform/README.md [hunk 9]
```diff
@@ -81,3 +80,34 @@ The following steps need to be followed:
...
+
+## Customizing the deployment further
+
+To see a full list of variables that can be changed to customize the deployment see:
+
+Infrastructure part: [https://github.com/sotec-iot/hono-extras/blob/terraform/terraform/infrastructure/variables.tf](https://github.com/sotec-iot/hono-extras/blob/terraform/terraform/software/variables.tf)<br/>
+Software part: [https://github.com/sotec-iot/hono-extras/blob/terraform/terraform/software/variables.tf](https://github.com/sotec-iot/hono-extras/blob/terraform/terraform/software/variables.tf)
+
+### Changing a variable
+To change the value of a variable, e.g. grafana_expose_externally, add the variable in the corresponding locals.tf file and assign it a new value.<br/>
```
Comment (julian-sotec) [inline terraform/README.md:92]:
```suggestion
To change the value of a variable, e.g. grafana_expose_externally, add the variable in the corresponding locals.tf file and assign a new value to it.<br/>
```

File: terraform/README.md [hunk 10]
```diff
@@ -81,3 +80,34 @@ The following steps need to be followed:
...
+```
+Replace `VARIABLE` with the name of the variable you want to change.<br/>
+For our example:
+```
+grafana_expose_externally = local.grafana_expose_externally
+```
+
+### Changing an object variable
+To change the properties within a variable of type object you have to follow the same steps as for a simple variable (described above) except you have to provide an object in the locals.tf file.
+For object variables that have optional properties you can just provide the non-optional properties and the ones you want to change.<br/>
```
Comment (julian-sotec) [inline terraform/README.md:109]:
```suggestion
For object variables that have optional properties, provide the non-optional properties and the ones you want to change.<br/>
```

File: terraform/README.md [hunk 11]
```diff
@@ -81,3 +80,34 @@ The following steps need to be followed:
...
+So for example the properties of `chart_version` and `algorithm` within the variable `enhanced_mqtt_load_balancer` can be change like seen below:
```
Comment (julian-sotec) [inline terraform/README.md:110]:
```suggestion
For example the properties of `chart_version` and `algorithm` within the variable `enhanced_mqtt_load_balancer` can be changed like this:
```

File: terraform/modules/hono/variables.tf [hunk 12]
```diff
@@ -207,3 +207,15 @@ variable "grafana_dns_name" {
...
+  type = object({
+    enabled       = bool
+    algorithm     = string
+  })
+  description = <<EOT
+Configuration options for the enhanced MQTT load balancer.
```
Comment (julian-sotec) [inline terraform/modules/hono/variables.tf:217]:
```suggestion
Configuration options for the advanced MQTT load balancer.
```

File: terraform/modules/hono/variables.tf [hunk 13]
```diff
@@ -207,3 +207,15 @@ variable "grafana_dns_name" {
...
+  enabled: Enables the use of the enhanced MQTT load balancer.
```
Comment (julian-sotec) [inline terraform/modules/hono/variables.tf:218]:
```suggestion
  enabled: Enables the use of the advanced MQTT load balancer.
```

File: terraform/modules/hono/variables.tf [hunk 14]
```diff
@@ -207,3 +207,15 @@ variable "grafana_dns_name" {
...
+  algorithm: Load balancing algorithm used by the enhanced MQTT load balancer. For a list of possible options see https://www.haproxy.com/documentation/kubernetes-ingress/community/configuration-reference/ingress/#load-balance .
```
Comment (julian-sotec) [inline terraform/modules/hono/variables.tf:219]:
```suggestion
  algorithm: Load balancing algorithm used by the advanced MQTT load balancer. For a list of possible options see https://www.haproxy.com/documentation/kubernetes-ingress/community/configuration-reference/ingress/#load-balance .
```

File: terraform/modules/hono/README.md [hunk 15]
```diff
@@ -34,6 +35,7 @@ No modules.
 | <a name="input_device_communication_static_ip_name"></a> [device\_communication\_static\_ip\_name](#input\_device\_communication\_static\_ip\_name) | Name of the Static IP for External Ingress | `string` | n/a | yes |
 | <a name="input_enable_http_adapter"></a> [enable\_http\_adapter](#input\_enable\_http\_adapter) | Used to enable the http adapter | `bool` | n/a | yes |
 | <a name="input_enable_mqtt_adapter"></a> [enable\_mqtt\_adapter](#input\_enable\_mqtt\_adapter) | Used to enable the mqtt adapter | `bool` | n/a | yes |
+| <a name="input_enhanced_mqtt_load_balancer"></a> [enhanced\_mqtt\_load\_balancer](#input\_enhanced\_mqtt\_load\_balancer) | Configuration options for the enhanced MQTT load balancer.<br>  enabled: Enables the use of the enhanced MQTT load balancer.<br>  algorithm: Load balancing algorithm used by the enhanced MQTT load balancer. For a list of possible options see https://www.haproxy.com/documentation/kubernetes-ingress/community/configuration-reference/ingress/#load-balance . | <pre>object({<br>    enabled       = bool<br>    algorithm     = string<br>  })</pre> | n/a | yes |
```
[the remark this answers is not in the document]
Comment (mattkaem) [reply terraform/modules/hono/README.md]: I did create a new mqtt_adapter object and added enhanced_mqtt_load_balancer (now named advanced_mqtt_load_balancer) object as a field. I would not change this readme as it is auto generated.

File: terraform/modules/hono/variables.tf (hunk 1 above)
Comment (mattkaem) [reply terraform/modules/hono/variables.tf]: done

File: terraform/modules/hono/README.md (hunk 2 above)
Comment (mattkaem) [reply terraform/modules/hono/README.md]: done

File: terraform/modules/hono/README.md (hunk 3 above)
Comment (mattkaem) [reply terraform/modules/hono/README.md]: done

File: terraform/modules/load_balancer/main.tf (hunk 4 above)
Comment (mattkaem) [reply terraform/modules/load_balancer/main.tf:14]: no, it will not

File: terraform/modules/load_balancer/README.md (hunk 5 above)
Comment (mattkaem) [reply terraform/modules/load_balancer/README.md]: I would prefer not to change anything here, as this readme is auto generated

File: terraform/software/README.md (hunk 6 above)
Comment (mattkaem) [reply terraform/software/README.md]: done

File: terraform/software/README.md (hunk 7 above)
Comment (mattkaem) [reply terraform/software/README.md]: done

File: terraform/software/README.md (hunk 8 above)
Comment (mattkaem) [reply terraform/software/README.md]: I would prefer not to change anything here, as this readme is auto generated

File: terraform/README.md (hunk 11 above)
Comment (mattkaem) [reply terraform/README.md]: Due to the change to the variable structure, this part is now completely changed.

File: terraform/modules/hono/variables.tf (hunk 12 above)
Comment (mattkaem) [reply terraform/modules/hono/variables.tf]: done

File: terraform/modules/hono/variables.tf (hunk 13 above)
Comment (mattkaem) [reply terraform/modules/hono/variables.tf]: done

File: terraform/modules/hono/variables.tf (hunk 14 above)
Comment (mattkaem) [reply terraform/modules/hono/variables.tf]: done

File: terraform/README.md (hunk 10 above)
Comment (mattkaem) [reply terraform/README.md]: Sentence changed to: "For object variables that have optional properties, it is sufficient to provide only the non-optional properties and the properties you want to change."

**Comment by julian-sotec** [comment]
LGTM!!

**Review by julian-sotec — APPROVED** [review]
LGTM!!

**Pull request merged**